subsys/bluetooth/host/gatt.c

```diff
@@ -5158,6 +5158,67 @@ int bt_gatt_store_ccc(uint8_t id, const bt_addr_le_t *addr)
 	return 0;
 }
 
+/* Struct used to store both the public and the random address of a device when replacing
+ * random addresses in the ccc attribute's cfg array with the device's public address after
+ * pairing complete.
+ */
+struct device_addresses {
+	const bt_addr_le_t *private_addr;
+	const bt_addr_le_t *public_addr;
+};
+
+uint8_t convert_to_public_on_match(const struct bt_gatt_attr *attr,
+				   uint16_t handle, void *user_data)
+{
+	struct _bt_gatt_ccc *ccc;
+	struct device_addresses *device_addresses = user_data;
+
+	/* Check if attribute is a CCC */
+	if (attr->write != bt_gatt_attr_write_ccc) {
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	ccc = attr->user_data;
+
+	/* Copy the device's public address to the config's address if the config's address is the
+	 * same as the device's private address
+	 */
+	for (size_t i = 0; i < ARRAY_SIZE(ccc->cfg); i++) {
+		if (bt_addr_le_cmp((const bt_addr_le_t *)&ccc->cfg[i].peer,
+		    device_addresses->private_addr) == 0) {
+			bt_addr_le_copy(&ccc->cfg[i].peer, device_addresses->public_addr);
+#ifndef CONFIG_BT_CTLR_ALLOW_SAME_PEER_CONN
```

Comment:
Host code cannot use controller kconfig options. On top of that, this is a hidden Kconfig option. Please remove the conditional compilation even though it will result in code a bit less optimized.